File: leetcode/go/_003/main.go
package main

import "fmt"

func main() {
	fmt.Println(lengthOfLongestSubstring("abcabcbb"))
	fmt.Println(lengthOfLongestSubstring("bbb"))
	fmt.Println(lengthOfLongestSubstring("pwwkew"))
}

func lengthOfLongestSubstring(s string) int {
	result := 0
	mp := make(map[byte]bool)

	l, r := 0, 0
	for r < len(s) {
		if !mp[s[r]] {
			mp[s[r]] = true
			result = max(result, r-l+1)
			r++
		} else {
			delete(mp, s[l])
			l++
		}
	}

	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}

	return b
}

File: leetcode/python/_003/main.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res, j, uniqs = 0, 0, set()
        for i in range(len(s)):
            while j < len(s) and s[j] not in uniqs:
                uniqs.add(s[j])
                j += 1
            res = max(res, j - i)
            uniqs.remove(s[i])
        return res
print(Solution().lengthOfLongestSubstring("abcabcbb"))